src/modules/mc_pos_control/MulticopterPositionControl.cpp

```diff
@@ -524,71 +552,71 @@ void MulticopterPositionControl::Run()
 			_takeoff_status_pub.update();
 		}
-
-		// save latest reset counters
-		_vxy_reset_counter = local_pos.vxy_reset_counter;
-		_vz_reset_counter = local_pos.vz_reset_counter;
-		_xy_reset_counter = local_pos.xy_reset_counter;
-		_z_reset_counter = local_pos.z_reset_counter;
-		_heading_reset_counter = local_pos.heading_reset_counter;
 	}
 
 	perf_end(_cycle_perf);
 }
 
-void MulticopterPositionControl::failsafe(const hrt_abstime &now, vehicle_local_position_setpoint_s &setpoint,
-		const PositionControlStates &states, bool warn)
+vehicle_local_position_setpoint_s MulticopterPositionControl::generateFailsafeSetpoint(const hrt_abstime &now,
+		const PositionControlStates &states)
 {
-	// Only react after a short delay
-	_failsafe_land_hysteresis.set_state_and_update(true, now);
+	// do not warn while we are disarmed, as we might not have valid setpoints yet
+	bool warn = _vehicle_control_mode.flag_armed && ((now - _last_warn) > 2_s);
 
-	if (_failsafe_land_hysteresis.get_state()) {
-		reset_setpoint_to_nan(setpoint);
-
-		if (PX4_ISFINITE(states.velocity(0)) && PX4_ISFINITE(states.velocity(1))) {
-			// don't move along xy
-			setpoint.vx = setpoint.vy = 0.f;
+	if (warn) {
+		PX4_WARN("invalid setpoints");
+		_last_warn = now;
+	}
 
-			if (warn) {
-				PX4_WARN("Failsafe: stop and wait");
-			}
+	vehicle_local_position_setpoint_s failsafe_setpoint{};
+	reset_setpoint_to_nan(failsafe_setpoint);
+	failsafe_setpoint.timestamp = now;
 
-		} else {
-			// descend with land speed since we can't stop
-			setpoint.acceleration[0] = setpoint.acceleration[1] = 0.f;
-			setpoint.vz = _param_mpc_land_speed.get();
+	if (PX4_ISFINITE(states.velocity(0)) && PX4_ISFINITE(states.velocity(1))) {
+		// don't move along xy
+		failsafe_setpoint.vx = failsafe_setpoint.vy = 0.f;
 
-			if (warn) {
-				PX4_WARN("Failsafe: blind land");
-			}
+		if (warn) {
+			PX4_WARN("Failsafe: stop and wait");
 		}
 
-		if (PX4_ISFINITE(states.velocity(2))) {
-			// don't move along z if we can stop in all dimensions
-			if (!PX4_ISFINITE(setpoint.vz)) {
-				setpoint.vz = 0.f;
-			}
+	} else {
+		// descend with land speed since we can't stop
+		failsafe_setpoint.acceleration[0] = failsafe_setpoint.acceleration[1] = 0.f;
+		failsafe_setpoint.vz = _param_mpc_land_speed.get();
 
-		} else {
-			// emergency descend with a bit below hover thrust
-			setpoint.vz = NAN;
-			setpoint.acceleration[2] = .3f;
+		if (warn) {
+			PX4_WARN("Failsafe: blind land");
+		}
+	}
 
-			if (warn) {
-				PX4_WARN("Failsafe: blind descend");
-			}
+	if (PX4_ISFINITE(states.velocity(2))) {
+		// don't move along z if we can stop in all dimensions
+		if (!PX4_ISFINITE(failsafe_setpoint.vz)) {
+			failsafe_setpoint.vz = 0.f;
 		}
 
-		_in_failsafe = true;
+	} else {
+		// emergency descend with a bit below hover thrust
+		failsafe_setpoint.vz = NAN;
+		failsafe_setpoint.acceleration[2] = .3f;
+
+		if (warn) {
+			PX4_WARN("Failsafe: blind descend");
```

Comment:
```suggestion
			PX4_WARN("Failsafe: blind descent");
```

Reply:
I was able to read over this typo over and over again 😄 
Fixed in 8c26730bd3f4f6f34159a5ac31e85cb5358a8711